x/superfluid/keeper/stake.go

```diff
@@ -656,3 +606,217 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
 		fn(index+int64(i), delegation)
 	}
 }
+
+// UnbondConvertAndStake converts given lock to osmo and stakes it to given validator.
+// Supports conversion of 1)superfluid bonded 2)superfluid undelegating 3)vanilla unlocking.
+// Liquid gamm shares will not be supported for conversion.
+// Delegation is done in the following logic:
+// - If valAddr provided, single delegate.
+// - If valAddr not provided and valset exists, valsetpref.Delegate
+// - If valAddr not provided and valset delegation is not possible, refer back to original lock's superfluid validator
+// - Else: error
+func (k Keeper) UnbondConvertAndStake(ctx sdk.Context, lockID uint64, sender, valAddr string,
+	minAmtToStake sdk.Int, sharesToConvert sdk.Coin) (totalAmtConverted sdk.Int, err error) {
+	senderAddr, err := sdk.AccAddressFromBech32(sender)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// use getMigrationType method to check status of lock (either superfluid staked, superfluid unbonding, vanilla locked, unlocked)
+	_, migrationType, err := k.getMigrationType(ctx, senderAddr, int64(lockID))
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// if superfluid bonded, first change it into superfluid undelegate to burn minted osmo and instantly undelegate.
+	if migrationType == SuperfluidBonded {
+		_, err = k.undelegateCommon(ctx, sender, lockID)
+		if err != nil {
+			return sdk.Int{}, err
+		}
+	}
+
+	if migrationType == SuperfluidBonded || migrationType == SuperfluidUnbonding || migrationType == NonSuperfluid {
+		totalAmtConverted, err = k.convertLockToStake(ctx, senderAddr, valAddr, lockID, minAmtToStake)
+	} else if migrationType == Unlocked {
+		totalAmtConverted, err = k.convertUnlockedToStake(ctx, senderAddr, valAddr, sharesToConvert, minAmtToStake)
+	} else { // liquid gamm shares without locks are not supported
+		return sdk.Int{}, fmt.Errorf("unsupported staking conversion type")
+	}
```

Comment:
I thought we agreed we were supporting this if the lockID is zero, which is why we added back sharesToConvert

Reply:
Forgot to update comment after adding back sharesToConvert, modified comment to match current status in https://github.com/osmosis-labs/osmosis/pull/5949/commits/18f4bd7e81d8dd238fc541fa4d9cd840ca82d0ed